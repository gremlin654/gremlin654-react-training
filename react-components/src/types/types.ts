export interface INotebook {
  id: string;
  model: string;
  img: string;
  screenDiagonal: string;
  screenResolution: string;
  screenRefreshRate: string;
  processorModel: string;
  RAM: string;
  SSDcapacity: string;
  VideoCardModel: string;
}

export interface IMainProps {
  notebooks: INotebook[];
}

export interface ICardProps {
  notebook: INotebook;
}

export type StateCardType = {
  like: string;
  likesNumber: number;
  favourite: string;
};

export interface IStateInput {
  inputValue: string;
}

export interface IPersonCard {
  name: string;
  surname: string;
  date: string;
  country: string;
  gender: boolean;
  img: string | null;
}
export interface IFormPageState {
  cards: IPersonCard[];
}
export interface IUserFormProps {
  onSubmit: (data: IPersonCard) => void;
}

export interface IUserFormState {
  errors: Record<string, unknown>;
  disabled: boolean;
}

export interface IUserCard {
  card: IPersonCard;
}

export interface IApiPageState {
  searchValue: string;
  isLoading: boolean;
  page: number;
  sortByStatus: string;
  characters: IСharacter[];
  isModalActive: boolean;
  modalData: ICharacterModalData;
}

export interface ICharacterModalData {
  id: number;
  name: string;
  species: string;
  gender: string;
  origin: string;
  image: string;
  created: string;
}

export interface ICharactersPagesInfo {
  count: number;
  pages: number;
  next: string | null;
  prev: string | null;
}
export interface IСharacter {
  id: number;
  name: string;
  status: string;
  species: string;
  type: string;
  gender: string;
  origin: {
    name: string;
    url: string;
  };
  location: {
    name: string;
    url: string;
  };
  image: string;
  episode: string[];
  url: string;
  created: string;
}
export interface IСharactersResponse {
  info: ICharactersPagesInfo;
  results: IСharacter[];
}

export interface IMyInputProps {
  searchValue: string;
  handleChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  clear: () => void;
  onKeyPressHandler: (event: React.KeyboardEvent<HTMLInputElement>) => void;
}

export interface IApiCardModalProps {
  handleChangeModalClose: () => void;
  modalData: ICharacterModalData;
}

export interface IApiCardProps {
  character: IСharacter;
  handleChangeCharacter: (character: ICharacterData) => void;
}

export interface IMySelectByStatusProps {
  onChangeStatus: (event: React.ChangeEvent<HTMLSelectElement>) => void;
  sortByStatus: string;
}
export interface IMySelectByGenderProps {
  onChangeGender: (event: React.ChangeEvent<HTMLSelectElement>) => void;
  sortByGender: string;
}
export interface IMySelectBySpeciesProps {
  onChangeSpecies: (event: React.ChangeEvent<HTMLSelectElement>) => void;
  sortBySpecies: string;
}

export type FormData = {
  name: string;
  surName: string;
  date: string;
  country: string;
  img: string;
  gender: boolean;
  agree: boolean;
};

export interface ICharacterData {
  id: number;
  name: string;
  species: string;
  gender: string;
  origin: string;
  image: string;
  created: string;
}

export interface IPaginationProps {
  setIsLoading: React.Dispatch<React.SetStateAction<boolean>>;
  getPages: (info: ICharactersPagesInfo) => void;
}

export type StateSearch = {
  searchValue: string;
  page: number;
  pages: number[];
  sortByStatus: string;
  sortByGender: string;
  sortBySpecies: string;
  characters: IСharacter[] | null;
  info: ICharactersPagesInfo | null;
  characterData: ICharacterData;
};

export type ActionSearch = {
  type: string;
  payload: StateSearch;
};

export type StateForm = {
  cards: IPersonCard[];
  formData: {
    name?: string | undefined;
    surName?: string | undefined;
    date?: string | undefined;
    country?: string | undefined;
    img?: File | undefined;
    gender?: boolean | undefined;
    agree?: boolean | undefined;
  };
};

export type ActionForm = {
  type: string;
  payload: StateForm;
};
