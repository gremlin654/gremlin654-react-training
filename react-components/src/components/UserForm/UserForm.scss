.user-form {
  &__container {
    display: flex;
    font-size: 1.8rem;
    flex-direction: column;
    min-width: 30rem;
    gap: 1.2rem;
  }
  &__input {
    border: 1px solid #00ffcc;
    font-size: 18px;
    height: 40px;
    margin-right: 0;
    max-width: 50rem;
    min-width: 30rem;
    padding: 0 0.5rem;
    &:focus-visible {
      outline: none;
    }
  }
  &__label {
    display: flex;
    flex-direction: column;
    align-items: baseline;
    align-content: space-between;
    justify-content: space-between;
  }
  &__label-checkbox-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
  }
  &__label-checkbox {
    display: flex;
    align-items: center;
    justify-content: center;
    column-gap: 0.5rem;
  }

  &__submit-container {
    display: flex;
    justify-content: center;
  }
  &__button {
    width: 14rem;
    height: 4rem;
    border: 0.1rem solid green;
    font-size: 2.2rem;
    -webkit-text-stroke-width: medium;

    &:hover {
      transition: 0.3s;
      background-color: green;
    }
  }
  &__file-uploader-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    gap: 1rem;
  }
  &__file-uploader-button {
    opacity: 0;
    position: absolute;
    z-index: -1;
  }
  &__file-uploader-button-costum {
    display: flex;
    cursor: pointer;
    background-color: yellow;
    border: 0.1rem solid;
    width: 14rem;
    height: 4rem;
    -webkit-text-stroke-width: medium;
    align-items: center;
    justify-content: center;
    &:hover {
      transition: 0.3s;
      background-color: rgb(200, 200, 8);
    }
  }
  &__preview-img {
    width: 20rem;
    height: 20rem;
    border: 0.1rem dashed gray;
    border-radius: 1.2rem;
  }
  &__errors {
    color: red;
    font-size: 1rem;
    height: 1.1rem;
  }
  &__switcher {
    display: flex;
    flex-direction: column;
    row-gap: 0.7rem;
    align-items: center;
  }
  &__switch-container {
    display: flex;
    column-gap: 0.7rem;
    align-items: center;
  }
  &__switch-text {
    font-size: 1.8rem;
    font-weight: 600;
  }

  &__switch-input {
    height: 0;
    width: 0;
    visibility: hidden;
  }

  &__switch-lable {
    cursor: pointer;
    text-indent: -9999px;
    width: 7rem;
    height: 2.5rem;
    background: #83ec97;
    display: block;
    border-radius: 10rem;
    position: relative;
    &:after {
      content: '';
      position: absolute;
      top: 0.11rem;
      left: 0.4rem;
      width: 2.25rem;
      height: 2.25rem;
      background: #fff;
      border-radius: 9rem;
      transition: 0.3s;
    }
  }

  &__switch-input:checked + &__switch-lable {
    background: #ec83b4;
  }

  &__switch-input:checked + &__switch-lable:after {
    left: calc(100% - 5px);
    transform: translateX(-100%);
  }
  &__switch-lable:active:after {
    width: 3rem;
  }
}
