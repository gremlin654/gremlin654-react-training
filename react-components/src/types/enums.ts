export enum SortTypeStatus {
  empty = '',
  alive = 'alive',
  dead = 'dead',
  unknown = 'unknown',
}

export enum SortTypeGender {
  empty = '',
  female = 'female',
  male = 'male',
  genderless = 'genderless',
  unknown = 'unknown',
}

export enum SortTypeSpecies {
  Empty = '',
  Alien = 'Alien',
  Human = 'Human',
  Animal = 'Animal',
  Poopybutthole = 'Poopybutthole',
  Humanoid = 'Humanoid',
  Cronenberg = 'Cronenberg',
  unknown = 'unknown',
}
