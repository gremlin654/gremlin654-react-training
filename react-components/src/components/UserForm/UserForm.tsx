import React, { ChangeEvent, useState, useEffect, useContext } from 'react';
import { IUserFormProps, FormData } from 'types/types';
import { FieldValues, SubmitHandler, useForm } from 'react-hook-form';
import { GlobalData } from 'context/context';
import { ACTION_TYPE } from 'types/enums';
import './UserForm.scss';

export const UserForm: React.FC<IUserFormProps> = ({ onSubmit }) => {
  const [photoPreview, setPhotoPreview] = useState<string>('./no_photo.jpg');
  const [disabledSubmit, setDisabledSubmit] = useState<boolean>(true);
  const { stateForm, dispatchForm } = useContext(GlobalData);
  const { agree, country, date, gender, name, surName } = stateForm.formData;

  const {
    register,
    handleSubmit,
    formState: { errors },
    reset,
    setValue,
    watch,
  } = useForm<FormData>({
    defaultValues: {
      name: '',
      surName: '',
      date: '',
      country: '',
      gender: false,
      agree: false,
    },
  });

  useEffect(() => {
    setValue('name', name as string);
    setValue('surName', surName as string);
    setValue('country', country as string);
    setValue('date', date as string);
    setValue('gender', gender as boolean);
    setValue('gender', gender as boolean);
    setValue('agree', agree as boolean);
    if (name || surName || country || date || agree) {
      setDisabledSubmit(false);
    }
  }, []);

  useEffect(() => {
    watch((data) => {
      dispatchForm({
        type: ACTION_TYPE.SET_FORM_DATA,
        payload: {
          ...stateForm,
          formData: {
            name: data.name,
            surName: data.surName,
            country: data.country,
            date: data.date,
            agree: data.agree,
            gender: data.gender,
          },
        },
      });
      if (data.agree || data.country || data.date || data.name || data.surName) {
        setDisabledSubmit(false);
      } else {
        setDisabledSubmit(true);
      }
    });
  }, [watch]);

  const handleUpload = (event: ChangeEvent<HTMLInputElement>) => {
    if (event.target.files?.length) {
      dispatchForm({
        type: ACTION_TYPE.SET_FORM_DATA,
        payload: {
          ...stateForm,
          formData: { ...stateForm.formData, img: event.target.files[0] },
        },
      });
      const fileLink = URL.createObjectURL((event.target.files as FileList)[0]);
      setPhotoPreview(fileLink);
    } else {
      setPhotoPreview('./no_photo.jpg');
    }
  };

  const onSubmitData: SubmitHandler<FieldValues> = (data) => {
    const formData = {
      name: data.name,
      surname: data.surName,
      date: data.date,
      country: data.country,
      img: photoPreview,
      gender: data.gender,
    };
    onSubmit(formData);
    reset();
    setPhotoPreview('./no_photo.jpg');
  };

  return (
    <form
      className="user-form__container"
      data-testid="user-form"
      onSubmit={handleSubmit((data) => onSubmitData(data))}
    >
      <label className="user-form__label">
        <span>Name:</span>
        <input
          {...register('name', {
            required: 'Please enter your first name',
            minLength: {
              value: 2,
              message: 'The first name must contain only letters',
            },
            pattern: {
              value: /^[a-zA-Za-яА-Я]+$/,
              message: 'The first name must contain only letters',
            },
          })}
          className={`user-form__input ${errors?.name && 'user-form__input-error'}`}
          data-testid="input-name"
        />
        <div className="user-form__errors">{errors?.name && errors.name.message}</div>
      </label>
      <label className="user-form__label">
        <span>Surname:</span>
        <input
          {...register('surName', {
            required: 'Please enter your last name',
            minLength: {
              value: 2,
              message: 'The last name must contain only letters',
            },
            pattern: {
              value: /^[a-zA-Za-яА-Я]+$/,
              message: 'The last name must contain only letters',
            },
          })}
          className={`user-form__input ${errors?.surName && 'user-form__input-error'}`}
          data-testid="input-surname"
        />
        <div className="user-form__errors">{errors?.surName && errors.surName.message}</div>
      </label>
      <label className="user-form__label" htmlFor="date">
        <span>Date of Birth:</span>
        <input
          {...register('date', {
            required: 'Please enter your birhtday',
          })}
          type="date"
          className={`user-form__input ${errors?.date && 'user-form__input-error'}`}
          data-testid="input-date"
        />
        <div className="user-form__errors">{errors?.date && errors.date.message}</div>
      </label>
      <label className="user-form__label">
        <span>What is your country:</span>
        <select
          {...register('country', {
            required: 'Please enter your country',
          })}
          className={`user-form__input ${errors?.country && 'user-form__input-error'}`}
          data-testid="select-country"
        >
          <option value="">Select...</option>
          <option value="belarus">Belarus</option>
          <option value="ukraine">Ukraine</option>
          <option value="poland">Poland</option>
          <option value="lithuania">Lithuania</option>
        </select>
        <div className="user-form__errors">{errors?.country && errors.country.message}</div>
      </label>
      <div className="user-form__switcher">
        <span>What is your gender?</span>
        <div className="user-form__switch-container">
          <span className="user-form__switch-text">Male</span>
          <input
            {...register('gender')}
            type="checkbox"
            id="switch"
            className="user-form__switch-input"
            data-testid="switch-gender"
          />
          <label htmlFor="switch" className="user-form__switch-lable">
            Toggle
          </label>
          <span className="user-form__switch-text">Female</span>
        </div>
      </div>
      <div className="user-form__file-uploader-container">
        Your avatar:
        <div className="user-form__file-input-container">
          <input
            {...register('img')}
            type="file"
            accept="image/jpeg,image/png"
            onChange={handleUpload}
            id="form-img"
            style={{ display: 'none' }}
            data-testid="input-file"
          />
          <label htmlFor="form-img" className="user-form__file-label">
            Change avatar
          </label>
          <img src={photoPreview} alt="preview" style={{ width: '7rem', height: '7rem' }} />
        </div>
      </div>
      <label className="user-form__label-checkbox-container">
        <div className="user-form__label-checkbox">
          <span>I consent to my personal data</span>
          <input
            {...register('agree', { required: 'Please you need agree' })}
            type="checkbox"
            data-testid="checkbox-personal"
          />
        </div>
        <div className="user-form__errors">{errors?.agree && errors.agree.message}</div>
      </label>
      <div className="user-form__submit-container">
        <button
          type="submit"
          className="user-form__button"
          data-testid="button-submit"
          disabled={disabledSubmit}
        >
          Submit
        </button>
      </div>
    </form>
  );
};
