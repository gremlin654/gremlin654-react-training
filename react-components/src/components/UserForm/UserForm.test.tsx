import { act, fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { IPersonCard } from '../../types/types';
import userEvent from '@testing-library/user-event';
import { UserForm } from './UserForm';

describe('UserForm component', () => {
  it('render UserForm', () => {
    render(
      <UserForm
        onSubmit={function (data: IPersonCard): void {
          {
            [data];
          }
        }}
      />
    );

    expect(screen.getByTestId('user-form')).toBeInTheDocument();
  });

  it('render name input', () => {
    render(
      <UserForm
        onSubmit={function (data: IPersonCard): void {
          {
            [data];
          }
        }}
      />
    );

    const inputName = screen.getByTestId('input-name');

    expect(inputName).toBeInTheDocument();

    userEvent.type(inputName, 'aaa');

    expect(inputName).toHaveValue('aaa');
  });

  it('render surname input', () => {
    render(
      <UserForm
        onSubmit={function (data: IPersonCard): void {
          {
            [data];
          }
        }}
      />
    );

    const inputSurname = screen.getByTestId('input-surname');

    expect(inputSurname).toBeInTheDocument();

    userEvent.type(inputSurname, 'aaa');

    expect(inputSurname).toHaveValue('aaa');
  });

  it('render data input', () => {
    render(
      <UserForm
        onSubmit={function (data: IPersonCard): void {
          {
            [data];
          }
        }}
      />
    );

    const inputData = screen.getByTestId('input-date');

    expect(inputData).toBeInTheDocument();

    userEvent.type(inputData, '2022-10-10');

    expect(inputData).toHaveValue('2022-10-10');
  });

  it('render file input', async () => {
    render(
      <UserForm
        onSubmit={function (data: IPersonCard): void {
          {
            [data];
          }
        }}
      />
    );

    const inputFile = screen.getByTestId('input-file') as HTMLInputElement;

    expect(inputFile).toBeInTheDocument();
  });

  it('render checkbox input', () => {
    render(
      <UserForm
        onSubmit={function (data: IPersonCard): void {
          {
            [data];
          }
        }}
      />
    );

    const inputCheckbox = screen.getByTestId('checkbox-personal');

    expect(inputCheckbox).toBeInTheDocument();

    userEvent.click(inputCheckbox);

    expect(inputCheckbox).toBeChecked();

    userEvent.click(inputCheckbox);

    expect(inputCheckbox).not.toBeChecked();
  });

  it('render switch input', () => {
    render(
      <UserForm
        onSubmit={function (data: IPersonCard): void {
          {
            [data];
          }
        }}
      />
    );

    const inputSwitch = screen.getByTestId('switch-gender');

    expect(inputSwitch).toBeInTheDocument();

    userEvent.click(inputSwitch);

    expect(inputSwitch).toBeChecked();

    userEvent.click(inputSwitch);

    expect(inputSwitch).not.toBeChecked();
  });

  it('render select ', () => {
    render(
      <UserForm
        onSubmit={function (data: IPersonCard): void {
          {
            [data];
          }
        }}
      />
    );

    const select = screen.getByTestId('select-country');
    const selectOption = screen.getByRole('option', { name: 'Ukraine' }) as HTMLOptionElement;

    expect(select).toBeInTheDocument();
    expect(selectOption).toBeInTheDocument();
    expect(screen.getAllByRole('option').length).toBe(5);

    userEvent.selectOptions(select, selectOption);

    expect(selectOption.selected).toBe(true);
    expect(select).toHaveValue('ukraine');
  });

  it('UserForm snaphot', () => {
    const userForm = render(
      <UserForm
        onSubmit={function (data: IPersonCard): void {
          {
            [data];
          }
        }}
      />
    );

    expect(userForm).toMatchSnapshot();
  });

  it('Submit is disabled', () => {
    render(
      <UserForm
        onSubmit={function (data: IPersonCard): void {
          {
            [data];
          }
        }}
      />
    );

    expect(screen.getByTestId('button-submit')).toBeInTheDocument();
    expect(screen.getByTestId('input-name')).toBeInTheDocument();
    expect(screen.getByTestId('button-submit')).toBeDisabled();

    userEvent.type(screen.getByTestId('input-name'), 'w');

    expect(screen.getByTestId('button-submit')).not.toBeDisabled();
  });

  it('to have errors if empty input and not check checkbox', async () => {
    render(
      <UserForm
        onSubmit={function (data: IPersonCard): void {
          {
            [data];
          }
        }}
      />
    );

    const inputName = screen.getByTestId('input-name');
    const inputSurname = screen.getByTestId('input-surname');
    const buttonSubmit = screen.getByTestId('button-submit');

    expect(inputName).toBeInTheDocument();
    expect(inputSurname).toBeInTheDocument();
    expect(buttonSubmit).toBeInTheDocument();
    expect(inputName).toHaveValue('');
    expect(inputSurname).toHaveValue('');
    expect(buttonSubmit).toBeDisabled();

    await act(async () => {
      fireEvent.input(inputSurname, {
        target: { value: 'wwww' },
      });
    });

    expect(inputSurname).toHaveValue('wwww');
    expect(buttonSubmit).not.toBeDisabled();
    await act(async () => {
      fireEvent.click(buttonSubmit);
    });

    expect(screen.queryByText(/Please enter your first name/i)).toBeInTheDocument();
    expect(screen.queryByText(/Please enter your last name/i)).not.toBeInTheDocument();
    expect(screen.queryByText(/Please enter your birhtday/i)).toBeInTheDocument();
    expect(screen.queryByText(/Please you need agree/i)).toBeInTheDocument();
  });
});
