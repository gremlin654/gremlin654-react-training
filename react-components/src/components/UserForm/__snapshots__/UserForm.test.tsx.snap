// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm component UserForm snaphot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="user-form__container"
        data-testid="user-form"
      >
        <label
          class="user-form__label"
        >
          <span>
            Name:
          </span>
          <input
            class="user-form__input undefined"
            data-testid="input-name"
            name="name"
          />
          <div
            class="user-form__errors"
          />
        </label>
        <label
          class="user-form__label"
        >
          <span>
            Surname:
          </span>
          <input
            class="user-form__input undefined"
            data-testid="input-surname"
            name="surName"
          />
          <div
            class="user-form__errors"
          />
        </label>
        <label
          class="user-form__label"
          for="date"
        >
          <span>
            Date of Birth:
          </span>
          <input
            class="user-form__input undefined"
            data-testid="input-date"
            name="date"
            type="date"
          />
          <div
            class="user-form__errors"
          />
        </label>
        <label
          class="user-form__label"
        >
          <span>
            What is your country:
          </span>
          <select
            class="user-form__input undefined"
            data-testid="select-country"
            name="country"
          >
            <option
              value=""
            >
              Select...
            </option>
            <option
              value="belarus"
            >
              Belarus
            </option>
            <option
              value="ukraine"
            >
              Ukraine
            </option>
            <option
              value="poland"
            >
              Poland
            </option>
            <option
              value="lithuania"
            >
              Lithuania
            </option>
          </select>
          <div
            class="user-form__errors"
          />
        </label>
        <div
          class="user-form__switcher"
        >
          <span>
            What is your gender?
          </span>
          <div
            class="user-form__switch-container"
          >
            <span
              class="user-form__switch-text"
            >
              Male
            </span>
            <input
              class="user-form__switch-input"
              data-testid="switch-gender"
              id="switch"
              name="gender"
              type="checkbox"
            />
            <label
              class="user-form__switch-lable"
              for="switch"
            >
              Toggle
            </label>
            <span
              class="user-form__switch-text"
            >
              Female
            </span>
          </div>
        </div>
        <div
          class="user-form__file-uploader-container"
        >
          Your avatar:
          <div
            class="user-form__file-input-container"
          >
            <input
              accept="image/jpeg,image/png"
              data-testid="input-file"
              id="form-img"
              name="img"
              style="display: none;"
              type="file"
            />
            <label
              class="user-form__file-label"
              for="form-img"
            >
              Change avatar
            </label>
            <img
              alt="preview"
              src="./no_photo.jpg"
              style="width: 7rem; height: 7rem;"
            />
          </div>
        </div>
        <label
          class="user-form__label-checkbox-container"
        >
          <div
            class="user-form__label-checkbox"
          >
            <span>
              I consent to my personal data
            </span>
            <input
              data-testid="checkbox-personal"
              name="agree"
              type="checkbox"
            />
          </div>
          <div
            class="user-form__errors"
          />
        </label>
        <div
          class="user-form__submit-container"
        >
          <button
            class="user-form__button"
            data-testid="button-submit"
            disabled=""
            type="submit"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="user-form__container"
      data-testid="user-form"
    >
      <label
        class="user-form__label"
      >
        <span>
          Name:
        </span>
        <input
          class="user-form__input undefined"
          data-testid="input-name"
          name="name"
        />
        <div
          class="user-form__errors"
        />
      </label>
      <label
        class="user-form__label"
      >
        <span>
          Surname:
        </span>
        <input
          class="user-form__input undefined"
          data-testid="input-surname"
          name="surName"
        />
        <div
          class="user-form__errors"
        />
      </label>
      <label
        class="user-form__label"
        for="date"
      >
        <span>
          Date of Birth:
        </span>
        <input
          class="user-form__input undefined"
          data-testid="input-date"
          name="date"
          type="date"
        />
        <div
          class="user-form__errors"
        />
      </label>
      <label
        class="user-form__label"
      >
        <span>
          What is your country:
        </span>
        <select
          class="user-form__input undefined"
          data-testid="select-country"
          name="country"
        >
          <option
            value=""
          >
            Select...
          </option>
          <option
            value="belarus"
          >
            Belarus
          </option>
          <option
            value="ukraine"
          >
            Ukraine
          </option>
          <option
            value="poland"
          >
            Poland
          </option>
          <option
            value="lithuania"
          >
            Lithuania
          </option>
        </select>
        <div
          class="user-form__errors"
        />
      </label>
      <div
        class="user-form__switcher"
      >
        <span>
          What is your gender?
        </span>
        <div
          class="user-form__switch-container"
        >
          <span
            class="user-form__switch-text"
          >
            Male
          </span>
          <input
            class="user-form__switch-input"
            data-testid="switch-gender"
            id="switch"
            name="gender"
            type="checkbox"
          />
          <label
            class="user-form__switch-lable"
            for="switch"
          >
            Toggle
          </label>
          <span
            class="user-form__switch-text"
          >
            Female
          </span>
        </div>
      </div>
      <div
        class="user-form__file-uploader-container"
      >
        Your avatar:
        <div
          class="user-form__file-input-container"
        >
          <input
            accept="image/jpeg,image/png"
            data-testid="input-file"
            id="form-img"
            name="img"
            style="display: none;"
            type="file"
          />
          <label
            class="user-form__file-label"
            for="form-img"
          >
            Change avatar
          </label>
          <img
            alt="preview"
            src="./no_photo.jpg"
            style="width: 7rem; height: 7rem;"
          />
        </div>
      </div>
      <label
        class="user-form__label-checkbox-container"
      >
        <div
          class="user-form__label-checkbox"
        >
          <span>
            I consent to my personal data
          </span>
          <input
            data-testid="checkbox-personal"
            name="agree"
            type="checkbox"
          />
        </div>
        <div
          class="user-form__errors"
        />
      </label>
      <div
        class="user-form__submit-container"
      >
        <button
          class="user-form__button"
          data-testid="button-submit"
          disabled=""
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
